ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/ExpectedRecordComparator.java

```diff
@@ -0,0 +1,157 @@
+
+public final class ExpectedRecordComparator {
+
+  private static final Map<JsonNodeType, BiPredicate<Object, JsonNode>> COMPARATORS =
+      ImmutableMap.<JsonNodeType, BiPredicate<Object, JsonNode>>builder()
+          .put(JsonNodeType.OBJECT, ExpectedRecordComparator::compareStruct)
+          .put(JsonNodeType.ARRAY, ExpectedRecordComparator::compareArray)
+          .put(JsonNodeType.NUMBER, ExpectedRecordComparator::compareNumber)
+          .put(JsonNodeType.STRING, ExpectedRecordComparator::compareText)
+          .put(JsonNodeType.BOOLEAN, ExpectedRecordComparator::compareBoolean)
+          .put(JsonNodeType.NULL, ExpectedRecordComparator::compareNull)
+          .build();
+
+  private ExpectedRecordComparator() {
+  }
+
+  public static boolean matches(final Object actualValue, final JsonNode expectedValue) {
+    return comparator(expectedValue).test(actualValue, expectedValue);
+  }
+
+  private static boolean compareStruct(final Object actualValue, final JsonNode expectedValue) {
+    final ObjectNode expected = (ObjectNode) expectedValue;
+    final Function<String, Object> getter;
+    if (actualValue instanceof Struct) {
+      getter = ((Struct) actualValue)::get;
+    } else if (actualValue instanceof Map) {
+      getter = ((Map<?, ?>) actualValue)::get;
+    } else {
+      return false;
+    }
+
+    final Iterator<Entry<String, JsonNode>> fields = expected.fields();
+    while (fields.hasNext()) {
+      final Entry<String, JsonNode> field = fields.next();
+      if (!comparator(field.getValue()).test(getter.apply(field.getKey()), field.getValue())) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private static boolean compareArray(final Object actualValue, final JsonNode expectedValue) {
+    final ArrayNode expected = (ArrayNode) expectedValue;
+    if (actualValue instanceof List) {
+      final List<?> actual = (List<?>) actualValue;
+      final Iterator<JsonNode> elements = expected.elements();
+
+      int i = 0;
+      while (elements.hasNext()) {
+        final JsonNode el = elements.next();
+        if (!comparator(el).test(actual.get(i), el)) {
+          return false;
+        }
+        i++;
+      }
+
+      return true;
+    }
+    return false;
+  }
+
+  private static boolean compareNumber(final Object actualValue, final JsonNode expectedValue) {
+    final NumericNode expected = (NumericNode) expectedValue;
+    if (actualValue instanceof Integer) {
+      return expected.intValue() == (Integer) actualValue;
+    }
+    if (actualValue instanceof Long) {
+      return expected.longValue() == (Long) actualValue;
+    }
+    if (actualValue instanceof Double) {
+      return expected.doubleValue() == (Double) actualValue;
+    }
+    if (actualValue instanceof BigDecimal) {
+      if (!expected.isBigDecimal()) {
+        // we don't want to risk comparing a BigDecimal with something of
+        // lower precision
+        return false;
+      }
+
+      expected.isBigDecimal();
```

Comment:
what does this line do?

Reply:
woops! I used IntelliJ "remove unused variable" and it kept the side-effect...